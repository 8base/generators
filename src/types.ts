import { TableSchema } from '@8base/utils';

export interface IGeneratorsCommonConfig {
  deep?: number;
  withMeta?: boolean;
}

export type GeneratorsConfig = {
  includeColumns?: string[],
} & IGeneratorsCommonConfig;

export interface IGeneratorsData {
  tablesList: TableSchema[];
  tableName: string;
  screenName?: string;
}

export interface IScreenTable {
  id?: string;
  tableName: string;
  screenName?: string;
  routeUrl?: string;
  tableFields?: string[];
  formFields?: string[];
}
