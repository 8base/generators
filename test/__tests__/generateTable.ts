

import { generateTable } from '../../src';
import { tablesSchema } from '../__fixtures__/tablesSchema';


it('should generate table by the table name', () => {
  const generatedTable = generateTable({
    tablesList: tablesSchema,
    tableId: 'PROPERTIES_ID',
  }, {
    deep: 2,
    withMeta: false,
  });

  expect(generatedTable).toMatchSnapshot();
});

