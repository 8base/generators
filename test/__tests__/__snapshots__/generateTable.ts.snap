// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should generate table by the table name 1`] = `
"import React from 'react';
import gql from 'graphql-tag';
import objectPath from 'object-path';
import { graphql, compose } from 'react-apollo';
import { DateTime } from 'luxon';
import { TableBuilder, Dropdown, Icon, Menu, Link, Tag, Row, withModal } from '@8base/boost';
import { FIELD_TYPE, SMART_FORMATS, FILE_FORMATS, DATE_FORMATS, SWITCH_FORMATS, SWITCH_VALUES } from '@8base/utils';

const PROPERTIES_LIST_QUERY = gql\`
query PropertiesTableContent($filter: PropertyFilter, $orderBy: [PropertyOrderBy], $after: String, $before: String, $first: Int, $last: Int, $skip: Int) {
  propertiesList(filter: $filter, orderBy: $orderBy, after: $after, before: $before, first: $first, last: $last, skip: $skip) {
    items {
      id
      address {
        country
      }
      phone {
        code
        number
      }
      pictures {
        items {
          id
          fileId
          filename
          downloadUrl
          shareUrl
          meta
        }
        count
      }
      bedrooms
      title
      description
      sqFootage
      bathrooms
      garage
      pool
      listings {
        items {
          id
          property {
            id
            _description
          }
          broker {
            id
            _description
          }
          buyer {
            id
            _description
          }
          seller {
            id
            _description
          }
          documents {
            items {
              id
              fileId
              filename
              downloadUrl
              shareUrl
              meta
            }
            count
          }
          status
          closingDate
          price
          _description
        }
        count
      }
      _description
    }
    count
  }
}
\`;

const TABLE_COLUMNS = [
  {
    name: 'address',
    title: 'Address',
    meta: {
      isList: false,
      fieldType: FIELD_TYPE.SMART,
      fieldTypeAttributes: {
        format: 'ADDRESS',
      },
    },
  },
  {
    name: 'phone',
    title: 'Phone',
    meta: {
      isList: false,
      fieldType: FIELD_TYPE.SMART,
      fieldTypeAttributes: {
        format: 'PHONE',
      },
    },
  },
  {
    name: 'pictures',
    title: 'Pictures',
    meta: {
      isList: true,
      fieldType: FIELD_TYPE.FILE,
      fieldTypeAttributes: {
        format: 'IMAGE',
      },
    },
  },
  {
    name: 'bedrooms',
    title: 'Bedrooms',
    meta: {
      isList: false,
      fieldType: FIELD_TYPE.NUMBER,
      fieldTypeAttributes: {
        format: 'NUMBER',
      },
    },
  },
  {
    name: 'title',
    title: 'Title',
    meta: {
      isList: false,
      fieldType: FIELD_TYPE.TEXT,
      fieldTypeAttributes: {
        format: 'UNFORMATTED',
      },
    },
  },
  {
    name: 'description',
    title: 'Description',
    meta: {
      isList: false,
      fieldType: FIELD_TYPE.TEXT,
      fieldTypeAttributes: {
        format: 'UNFORMATTED',
      },
    },
  },
  {
    name: 'sqFootage',
    title: 'SqFootage',
    meta: {
      isList: false,
      fieldType: FIELD_TYPE.NUMBER,
      fieldTypeAttributes: {
        format: 'NUMBER',
      },
    },
  },
  {
    name: 'bathrooms',
    title: 'Bathrooms',
    meta: {
      isList: false,
      fieldType: FIELD_TYPE.NUMBER,
      fieldTypeAttributes: {
        format: 'NUMBER',
      },
    },
  },
  {
    name: 'garage',
    title: 'Garage',
    meta: {
      isList: false,
      fieldType: FIELD_TYPE.SWITCH,
      fieldTypeAttributes: {
        format: 'YES_NO',
      },
    },
  },
  {
    name: 'pool',
    title: 'Pool',
    meta: {
      isList: false,
      fieldType: FIELD_TYPE.SWITCH,
      fieldTypeAttributes: {
        format: 'YES_NO',
      },
    },
  },
  {
    name: 'listings',
    title: 'Listings',
    meta: {
      isList: true,
      fieldType: FIELD_TYPE.RELATION,
      fieldTypeAttributes: {
        format: '',
      },
    },
  },
  {
    name: 'edit',
    title: '',
    width: '60px',
  },
];

const enhancer = compose(
  withModal,
  graphql(PROPERTIES_LIST_QUERY, { name: 'properties' })
);

const PropertiesTable = enhancer(
  class PropertiesTable extends React.PureComponent {
    renderEdit = rowData => (
      <Dropdown defaultOpen={false}>
        <Dropdown.Head>
          <Icon name=\\"More\\" size=\\"sm\\" color=\\"LIGHT_GRAY2\\" />
        </Dropdown.Head>
        <Dropdown.Body pin=\\"right\\">
          {({ closeDropdown }) => (
            <Menu>
              <Menu.Item
                onClick={() => {
                  this.props.openModal('PROPERTY_EDIT_DIALOG_ID', { id: rowData.id });
                  closeDropdown();
                }}
              >
                Edit
              </Menu.Item>
              <Menu.Item
                onClick={() => {
                  this.props.openModal('PROPERTY_DELETE_DIALOG_ID', { id: rowData.id });
                  closeDropdown();
                }}
              >
                Delete
              </Menu.Item>
            </Menu>
          )}
        </Dropdown.Body>
      </Dropdown>
    );

    renderItems = (column, rowData, handler) => {
      const dataPath = column.name.split('.');
      const cellData = objectPath.get(rowData, dataPath) || '';

      if (column.meta.isList) {
        const itemsArray = cellData.items ? cellData.items : cellData;

        return (
          <Row style={{ flexWrap: 'wrap' }}>
            {itemsArray && itemsArray.map(item => !!item && <Tag color=\\"LIGHT_GRAY2\\">{handler(item)}</Tag>)}
          </Row>
        );
      } else {
        return cellData && <div>{handler(cellData)}</div>;
      }
    };

    renderScalar = (column, rowData) => {
      return this.renderItems(column, rowData, item => item);
    };

    renderDate = (column, rowData) => {
      const dateFormat =
        column.meta.fieldTypeAttributes.format === DATE_FORMATS.DATE ? DateTime.DATE_SHORT : DateTime.DATETIME_SHORT;

      return this.renderItems(column, rowData, item => DateTime.fromISO(item).toLocaleString(dateFormat));
    };

    renderSwitch = (column, rowData) => {
      if (column.meta.fieldTypeAttributes.format === SWITCH_FORMATS.CUSTOM) {
        return this.renderItems(column, rowData, item => item);
      } else {
        return this.renderItems(column, rowData, item => SWITCH_VALUES[column.meta.fieldTypeAttributes.format][item]);
      }
    };

    renderRelation = (column, rowData) => {
      const dataPath = column.name.split('.');

      if (column.meta.isList) {
        return objectPath.get(rowData, [...dataPath, 'count']) || '';
      } else {
        return objectPath.get(rowData, [...dataPath, '_description']) || '';
      }
    };

    renderFile = (column, rowData) => {
      if (column.meta.fieldTypeAttributes.format === FILE_FORMATS.IMAGE && !column.meta.isList) {
        const cellData = objectPath.get(rowData, column.name.split('.'));

        return cellData && <img src={cellData.downloadUrl} alt=\\"filename\\" style={{ width: '5rem' }} />;
      } else {
        return this.renderItems(column, rowData, item => (
          <div>
            <Link key={item.downloadUrl} target=\\"_blank\\" href={item.downloadUrl} size=\\"sm\\">
              {item.filename}
            </Link>
          </div>
        ));
      }
    };

    renderSmart = (column, rowData) => {
      return this.renderItems(column, rowData, item => {
        if (column.meta.fieldTypeAttributes.format === SMART_FORMATS.ADDRESS) {
          const address = [item.street1, item.street2, item.city, item.state, item.zip].filter(item => !!item);

          return join(', ', address);
        } else if (column.meta.fieldTypeAttributes.format === SMART_FORMATS.PHONE) {
          const phone = [item.code, item.number].filter(item => !!item);

          return join(' ', phone);
        }
      });
    };

    renderCell = (column, rowData) => {
      if (column.name === 'edit') {
        return this.renderEdit(rowData);
      }

      switch (column.meta.fieldType) {
        case FIELD_TYPE.TEXT:
        case FIELD_TYPE.NUMBER:
          return this.renderScalar(column, rowData);

        case FIELD_TYPE.DATE:
          return this.renderDate(column, rowData);

        case FIELD_TYPE.SWITCH:
          return this.renderSwitch(column, rowData);

        case FIELD_TYPE.FILE:
          return this.renderFile(column, rowData);

        case FIELD_TYPE.RELATION:
          return this.renderRelation(column, rowData);

        case FIELD_TYPE.SMART:
          return this.renderSmart(column, rowData);

        default:
          return null;
      }
    };

    openCreateModal = () => {
      const { openModal } = this.props;

      openModal('PROPERTY_CREATE_DIALOG_ID');
    };

    render() {
      const { properties } = this.props;
      const tableData = objectPath.get(properties, ['propertiesList', 'items']) || [];

      return (
        <TableBuilder
          loading={properties.loading}
          data={tableData}
          columns={TABLE_COLUMNS}
          action=\\"Create Property\\"
          renderCell={this.renderCell}
          onActionClick={this.openCreateModal}
        />
      );
    }
  }
);

export { PropertiesTable };
"
`;
